client/src/components/UsedOnTable/provideUsedOnData.js

```diff
@@ -1,51 +1,47 @@
-import React, { Component } from 'react';
+import React, { useEffect, useState } from 'react';
 import PropTypes from 'prop-types';
 import { connect } from 'react-redux';
 import { loadUsedOn } from 'state/usedOn/usedOnActions';
 import { injectTabContext } from 'hooks/useTabContext';
 
 const provideUsedOnData = (UsedOnTable) => {
-  class UsedOnDataProvider extends Component {
-    componentDidMount() {
-      const firstMount = this.haveFetchedData === undefined;
-      const tabContext = this.props.tabContext ? this.props.tabContext : false;
-      const activeTab = tabContext ? tabContext.isOnActiveTab : false;
-      this.haveFetchedData = false;
-      if ((firstMount && activeTab) || this.props.forceFetch) {
-        this.fetchDataFromEndpoint();
-      }
-    }
+  const UsedOnDataProvider = ({
+    identifier,
+    loading,
+    data,
+    error,
+    usedOn,
+    forceFetch,
+    tabContext,
+    loadUsedOn: loadUsedOnAction,
+  }) => {
+    const [haveFetchedData, setHaveFetchedData] = useState(false);
 
-    componentDidUpdate(oldProps) {
-      const tabContext = this.props.tabContext;
+    const fetchDataFromEndpoint = () => {
+      const { method, url } = (data.readUsageEndpoint || {});
+      if (!haveFetchedData || forceFetch) {
+        // see client/src/state/usedOn/usedOnActions.js
+        loadUsedOnAction(identifier, method, url);
+      }
+      setHaveFetchedData(true);
+    };
 
-      // Fetch data if
+    useEffect(() => {
+      // Fetch data if on of the following is true:
       // - identifier has changed
       // - component is not in a tab
       // - component is on the currently active tab
-      const doFetch =
-        oldProps.identifier !== this.props.identifier ||
-        !tabContext ||
-        tabContext.isOnActiveTab;
-
-      if (doFetch) {
-        this.fetchDataFromEndpoint();
+      if (!tabContext || (tabContext && tabContext.isOnActiveTab)) {
+        fetchDataFromEndpoint();
```

Comment:
The `componentDidMount()` logic used to be based partially on the value of `haveFetchedData` being `undefined`. We explicitly only called `fetchDataFromEndpoint()` if `forceFetch` was true or if this was the "first" mount.

This will now fetch data if `tabContext` is falsy, or if `tabContext` and `tabContext.isOnActiveTab` are truthy - which seems like a different condition to me and doesn't even include `forceFetch` in the condition. 

Should probably be like this?

```suggestion
      if ((haveFetchedData !== undefined && tabContext?.isOnActiveTab) || forceFetch) {
        fetchDataFromEndpoint();
```

And then for the `useState` have

```diff
-    const [haveFetchedData, setHaveFetchedData] = useState(false);
+    const [haveFetchedData, setHaveFetchedData] = useState(undefined);
```

Reply:
I don't understand why we would use `undefined` for something that's a bool?

The old code `const firstMount = this.haveFetchedData === undefined;` just seems like it was an anti-pattern. I ended up kind of manually refactoring a fair bit of this HOC cos it was kind of weirdly put together

I've manually tested that it works correctly:
- in its usual context where it's the 3rd tab in the file detail view
- it only makes the AJAX request when you click on the tab
- it will correctly only make a single ajax request when clicking the 2nd tab and then the 3rd tab again
- in a non-tab context by commenting out [these 2 lines](https://github.com/silverstripe/silverstripe-asset-admin/blob/3/code/Forms/FileFormFactory.php#L60-L61)

Reply:
> I don't understand why we would use undefined for something that's a bool?

Doesn't matter since you've pointed out before that this is replacing `componentDidUpdate` not `componentDidMount`

Reply:
> I ended up kind of manually refactoring a fair bit of this HOC cos it was kind of weirdly put together

I was expecting a direct like-for-like - when you refactor things in these class-to-functional-component PRs _please_ point out every change that isn't a straight like-for-like move before marking the PR ready for review.